### DISCUSSION TITLE
ODataServiceFactory undefined error

### DISCUSSION THREAD

**Discussion by AnaisR7** [awdorrin/ngx-odata-client#1] [Q&A] [answered]
Hi @awdorrin !
Thank you for promptly updating and publishing your work on npm! I appreciate your quick response.

I'm encountering some issues while attempting to update my project to use your package, and need some help.

I've uninstalled angular-odata-es5, installed your package (@awdorrin/ngx-odata-client) from npm, and updated my imports accordingly.

I also added an index.ts file in the @awdorrin/ngx-odata-client folder to simplify imports by grouping them together (for example: `import { NgxODataClientModule, ODataConfiguration, ODataServiceFactory } from '@awdorrin/ngx-odata-client';`).

Although there are no compilation errors and the project builds successfully, I'm encountering an error in my browser upon launch:
`Error: Invalid provider for the NgModule 'AppModule' - only instances of Provider and Type are allowed, got: [..., ..., ..., ..., ?undefined?, ..., ..., ..., ..., ..., ..., ..., ...]`

In the providers array of my AppModule, I haven't made any changes from what I used with the previous package (except for the imports path):
```
{
      // OData Configuration
      provide: ODataConfiguration,
      useClass: ODataConfigurationFactory  // personnal class that overload the ODataConfiguration baseUrl and defaultRequestOptions and customRequestOptions headers 
  },
  ODataServiceFactory,
  ```
  
The "?undefined?" refers to ODataServiceFactory.

I've tried various solutions without success. I also attempted removing this code from the providers section and adding NgxODataClientModule to the imports section, but encountered a different error.

I am on angular 15.2.10 and using v15.0.0 of your package. Tried with angular 16 and your 16.0.0 also, same "?undefined?" on ODataServiceFactory issue.

Any assistance you could provide would be greatly appreciated.

Thank you!

**Comment by awdorrin** [comment]
Here is what I do in my application:

In app.module.ts, make sure you have the following:

```
import { ODataConfiguration, ODataServiceFactory } from "@awdorrin/ngx-odata-client";

@NgModule({
...
providers: [
 ...,
   ODataServiceFactory,
   { provide: ODataConfiguration, useClass: ODataConfigService },
 ...]
```

ODataConfigService is: service/odata-config-service.ts:
```
import { Injectable, Inject } from '@angular/core';
import { ODataConfiguration } from "@awdorrin/ngx-odata-client";

@Injectable({
  providedIn: 'root'
})
export class ODataConfigService extends ODataConfiguration {
  constructor(
    @Inject('BASE_URL') private host: string
  ) {
    super();
    this.baseUrl = this.host + "odata/";
  }
}
```
Each of my components where I replaced angular-odata-es5 with my version:
```
//import { ODataServiceFactory, ODataService } from "angular-odata-es5";
import { ODataServiceFactory, ODataService } from "@awdorrin/ngx-odata-client";

// everything else stayed the same...

export class ActionService implements OnInit {
  private action: ActionItem;
  private user: UserProfile;
  private odataActionItem: ODataService<ActionItem>;

  //---------------------------------------------------------------------------
  //
  //---------------------------------------------------------------------------
  constructor(
    private route: ActivatedRoute,
    private odataFactory: ODataServiceFactory,
    private userService: UserProfileService
  ) {

```

**Comment by AnaisR7** [comment]
Still haven't succeeded in making it work. I encountered a new error when changing my **ODataConfigurationFactory** to your **ODataConfigService** in the **useClass** property. The error is: `Uncaught TypeError: class heritage @awdorrin/ngx-odata-client__WEBPACK_IMPORTED_MODULE_0__.ODataConfiguration is not an object or null`.

I have created a very basic Angular (v15.2.10) project that reproduces the error I encountered with your provided code. You can find it here: https://github.com/AnaisR7/stackblitz15/tree/main/src/app.

In AppModule you can switch the provider with my ODataConfigurationFactory that generate the initial error **?undefined?** I had on ODataServiceFactory provider.

**Reply by awdorrin** [reply]
Taking a look, I think I see where I made a mistake with the configuration for deploying to npm.
Will post back once I get it working.

**Reply by awdorrin** [reply, marked as answer]
I have published @awdorrin/ngx-odata-client@15.0.2 which is packaged properly.
If you give me permission to your stackblitz15, I will push my updates that resolve the build issue.
I'll work on updates for Angular 16 and 17 and get them published some time today.

**Reply by awdorrin** [reply]
16.0.1 and 17.01 have been published for angular 16 and 17, respectively.

**Reply by awdorrin** [reply]
Currently, linq-collections package has to be installed in addition to the @awdorrin/ngx-odata-client. 
I think I need to move it from peerDependencies to dependencies, so it will be installed automatically, but will have to experiment with that at a later time.

**Reply by AnaisR7** [reply]
Certainly, I've granted you the permissions. This will serve as an example for me to implement in my project. 
Thank you for your responsiveness !

**Reply by awdorrin** [reply]
You caught me on a slow work day :-)
I just pushed up to your stackblitz15 repo. Hopefully that works for you, if not, let me know.

**Reply by AnaisR7** [reply]
Perfect :) 
I pulled. The test project stackblitz15 run without any errors. So I made the necessary changes to my project and all looks good also! Now, I'll be able to proceed with upgrading to Angular versions 16 and 17.
